panic: runtime error: invalid memory address or nil pointer dereference

### Terraform Version

```shell
Terraform v1.10.2
on windows_amd64
+ provider registry.opentofu.org/hashicorp/aws v5.80.0
+ provider registry.opentofu.org/hashicorp/random v3.6.3
+ provider registry.opentofu.org/hashicorp/tls v4.0.6
+ provider registry.terraform.io/hashicorp/aws v5.16.2
+ provider registry.terraform.io/hashicorp/random v3.6.2
+ provider registry.terraform.io/hashicorp/tls v4.0.5
```


### Terraform Configuration Files

```
[us-east-1.zip](https://github.com/user-attachments/files/18116999/us-east-1.zip)
```


### Debug Output

```
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

panic: runtime error: invalid memory address or nil pointer dereference
goroutine 571 [running]:
runtime/debug.Stack()
        runtime/debug/stack.go:26 +0x5e
github.com/hashicorp/terraform/internal/logging.PanicHandler()
        github.com/hashicorp/terraform/internal/logging/panic.go:84 +0x18a
panic({0x319c660?, 0x5945440?})
        runtime/panic.go:785 +0x132
github.com/hashicorp/terraform/internal/plans.(*ChangesSrc).Empty(0x0)
        github.com/hashicorp/terraform/internal/plans/changes_src.go:43 +0x25
github.com/hashicorp/terraform/internal/terraform.(*Context).planWalk(0xc001080680, 0xc000aa40e0, 0xc000143a40, 0xc000f946e0)
        github.com/hashicorp/terraform/internal/terraform/context_plan.go:838 +0x13ef
github.com/hashicorp/terraform/internal/terraform.(*Context).plan(0x0?, 0x0?, 0x0?, 0xc000915dd0?)
        github.com/hashicorp/terraform/internal/terraform/context_plan.go:418 +0x1c
github.com/hashicorp/terraform/internal/terraform.(*Context).PlanAndEval(0xc001080680, 0xc000aa40e0, 0xc000143a40, 0xc000f946e0)
        github.com/hashicorp/terraform/internal/terraform/context_plan.go:274 +0xa57
github.com/hashicorp/terraform/internal/terraform.(*Context).Plan(...)
        github.com/hashicorp/terraform/internal/terraform/context_plan.go:156
github.com/hashicorp/terraform/internal/backend/local.(*Local).opPlan.func2()
        github.com/hashicorp/terraform/internal/backend/local/backend_plan.go:108 +0xe5
created by github.com/hashicorp/terraform/internal/backend/local.(*Local).opPlan in goroutine 79
        github.com/hashicorp/terraform/internal/backend/local/backend_plan.go:104 +0x79a
```


### Expected Behavior

output from standard "terraform plan" command

### Actual Behavior

terraform plan output reaches a point and then crashes

### Steps to Reproduce

1. terraform init
2. terraform plan

### Additional Context

_No response_

### References

_No response_

### Generative AI / LLM assisted development?

_No response_

--- a/internal/plans/changes.go
+++ b/internal/plans/changes.go
@@ -40,12 +40,14 @@ func NewChanges() *Changes {
 
 // Encode encodes all the stored resource and output changes into a new *ChangeSrc value
 func (c *Changes) Encode(schemas *schemarepo.Schemas) (*ChangesSrc, error) {
+	// a plan is always built even when there are errors, so make sure to return
+	// a valid changesSrc.
 	changesSrc := NewChangesSrc()
 
 	for _, rc := range c.Resources {
 		p, ok := schemas.Providers[rc.ProviderAddr.Provider]
 		if !ok {
-			return nil, fmt.Errorf("Changes.Encode: missing provider %s for %s", rc.ProviderAddr, rc.Addr)
+			return changesSrc, fmt.Errorf("Changes.Encode: missing provider %s for %s", rc.ProviderAddr, rc.Addr)
 		}
 
 		var schema providers.Schema
@@ -59,12 +61,12 @@ func (c *Changes) Encode(schemas *schemarepo.Schemas) (*ChangesSrc, error) {
 		}
 
 		if schema.Block == nil {
-			return nil, fmt.Errorf("Changes.Encode: missing schema for %s", rc.Addr)
+			return changesSrc, fmt.Errorf("Changes.Encode: missing schema for %s", rc.Addr)
 		}
 
 		rcs, err := rc.Encode(schema.Block.ImpliedType())
 		if err != nil {
-			return nil, fmt.Errorf("Changes.Encode: %w", err)
+			return changesSrc, fmt.Errorf("Changes.Encode: %w", err)
 		}
 
 		changesSrc.Resources = append(changesSrc.Resources, rcs)
@@ -73,7 +75,7 @@ func (c *Changes) Encode(schemas *schemarepo.Schemas) (*ChangesSrc, error) {
 	for _, ocs := range c.Outputs {
 		oc, err := ocs.Encode()
 		if err != nil {
-			return nil, err
+			return changesSrc, err
 		}
 		changesSrc.Outputs = append(changesSrc.Outputs, oc)
 	}

--- a/internal/plans/changes_test.go
+++ b/internal/plans/changes_test.go
@@ -43,3 +43,27 @@ func TestChangeEncodeSensitive(t *testing.T) {
 		})
 	}
 }
+
+// make sure we get a valid value back even when faced with an error
+func TestChangeEncodeError(t *testing.T) {
+	changes := &Changes{
+		Outputs: []*OutputChange{
+			{
+				// Missing Addr
+				Change: Change{
+					Before: cty.NullVal(cty.DynamicPseudoType),
+					// can't encode a marked value
+					After: cty.StringVal("test").Mark("shoult not be here"),
+				},
+			},
+		},
+	}
+	// no resources so we can get by with no schemas
+	changesSrc, err := changes.Encode(nil)
+	if err == nil {
+		t.Fatal("expected error")
+	}
+	if changesSrc == nil {
+		t.Fatal("changesSrc should not be nil")
+	}
+}

